compiler/parser/RdrHsSyn.hs

```diff
@@ -2304,127 +2384,97 @@ classInstDecl tyApps funBinds =
-      loc = getLoc body
-      arg = case argPatType of
-               -- See note [Controlled choice groups - Old syntax]
-               ArgWildcardPat -> WildPat noExt
-               ArgAsPat -> argPatOfChoice choiceConName mbChoiceFields
-      ctx = matchContext tag
-      match = matchWithBinds ctx [this, arg] loc body binds
-      match_group = matchGroup loc match
-  funBind loc tag match_group
-
--- | Construct a 'consuming' function binding.  This will only return
--- @Just@ a function if the provied argument is 'Just style'
--- indicating a choice is "postconsuming" or "nonconsuming".
-mkTemplateConsumingFunBindDecl
-  :: Located (Maybe String)
-  -> P (Maybe (LHsBind GhcPs))
-mkTemplateConsumingFunBindDecl (L loc consuming) =
-  case consuming of
-    Nothing -> return Nothing
-    Just style -> do
-      let tag = noLoc $ mkRdrUnqual (mkVarOcc "consuming")
-          ctx = matchContext tag
-          body = L loc $ HsVar noExt $ L loc $ qualifyDesugar $ mkVarOcc style
-          match = matchWithBinds ctx [] loc body Nothing
-          match_group = matchGroup loc match
-      funBind loc tag match_group
-
--- | Construct a 'choice' function binding.
-mkTemplateChoiceFunBindDecl ::
-     Located RdrName                        -- data ctor 'T'
-  -> Located RdrName                        -- data ctor 'S'
-  -> Maybe (LHsType GhcPs)                  -- maybe record fields
-  -> Located ([AddAnn], [ExprLStmt GhcPs])  -- body is a @do@ stmt list
-  -> Maybe (LHsLocalBinds GhcPs)            -- local binds
-  -> P (Maybe (LHsBind GhcPs))              -- function binding
-mkTemplateChoiceFunBindDecl
-  conName choiceConName mbChoiceFields body@(L loc (_ann, stmts)) binds = do
-  let tag = noLoc $ mkRdrUnqual (mkVarOcc "choice")
-      this = asPatRecWild "this" conName
-      self = XPat (noLoc $ VarPat noExt $ noLoc $ mkRdrUnqual (mkVarOcc "self"))
-      arg = argPatOfChoice choiceConName mbChoiceFields
-      do_ = L loc $ HsDo noExt DoExpr (L loc stmts)
-      ctx = matchContext tag
-      match = matchWithBinds ctx [this, self, arg] loc do_ binds
-      match_group = matchGroup loc match
-  funBind (combineLocs choiceConName body) tag match_group
+      choiceNameToRdrName = if choiceName == "Archive" then qualifyDesugar else mkRdrUnqual
+      arg = argPatOfChoice (noLoc $ choiceNameToRdrName $ mkDataOcc choiceName) cdChoiceFields
+      controllerArg = if flexible then arg else WildPat noExt
+      mkMethod :: String -> [Pat GhcPs] -> Bool -> LHsExpr GhcPs -> LHsBind GhcPs
+      mkMethod methodName args includeBindings body =
+        let fullMethodName = prefixTemplateClassMethod $ methodName ++ templateName ++ choiceName in
+        mkTemplateClassMethod fullMethodName args body $
+          -- General rule: only include template bindings for methods with `this` in scope
+          if includeBindings then Just binds else Nothing
+      mkVarQual :: OccName -> HsExpr GhcPs = HsVar noExt . noLoc . qualifyDesugar
```

Comment:
Definitely seems to be a lot of repetition of this. See above about lifting `mkVarQual` to file scope.